src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -579,5 +581,60 @@ public void ChainHasOneCheckPointAndAssumeValid_ChainsWithCheckpointButMissedAss
         }
+
+        /// <summary>
+        /// Issue 17 @ We advanced consensus tip (CT) and there are some partially validated (PV) headers after the CT.
+        /// Now we receive headers that are after the last PV header and include assume valid (AV). Make sure that those
+        /// headers that are before the AV header and after the last PV are all marked as AV. 
+        /// </summary>
+        [Fact]
+        public void ChainHasPartiallyValidatedAfterConsensusTip_NewHeadersWithAssumeValidPresented_CorrectHeadersAreMarkedAsAssumedValid()
+        {
+            // Chain header tree setup.
+            // Initial chain has 4 header with the consensus tip at header 4.
+            // Example: fv1=fv2=fv3=fv4 (fv - fully validated).
+            const int initialChainSize = 4;
+            TestContext ctx = new TestContextBuilder().WithInitialChain(initialChainSize).Build();
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader initialChainTip = ctx.InitialChainTip;
+
+            // Extend the chain with 2 partially validated headers.
+            // Example: fv1=fv2=fv3=(fv4)=pv5=pv6 (pv - partially validated).
+            const int partiallyValidatedHeadersCount = 2;
+            initialChainTip = ctx.ExtendAChain(partiallyValidatedHeadersCount, initialChainTip);
+            
+            // Chain is presented by peer 1.
+            // Mark pv5 and pv6 as partially validated.
+            List<BlockHeader> listOfCurrentChainHeaders =
+                ctx.ChainedHeaderToList(initialChainTip, partiallyValidatedHeadersCount);
+            ConnectNewHeadersResult result = cht.ConnectNewHeaders(1, listOfCurrentChainHeaders);
+            initialChainTip = result.Consumed;
+            initialChainTip.BlockValidationState = ValidationState.PartiallyValidated;
+            initialChainTip.Previous.BlockValidationState = ValidationState.PartiallyValidated;
+
+            // Extend the chain with 6 normal headers, where header at the height 9 is an "assumed valid" header.
+            // Example: fv1=fv2=fv3=(fv4)=pv5=pv6=h7=h8=(av9)=h10=h11=h12 (av - assumed valid).
+            const int extensionHeadersCount = 6;
+            initialChainTip = ctx.ExtendAChain(extensionHeadersCount, initialChainTip);
+            ChainedHeader assumedValidHeader = initialChainTip.GetAncestor(9);
+            assumedValidHeader.BlockValidationState = ValidationState.AssumedValid; 
+            ctx.ConsensusSettings.BlockAssumedValid = assumedValidHeader.HashBlock;
+            listOfCurrentChainHeaders = 
+                    ctx.ChainedHeaderToList(initialChainTip, extensionHeadersCount);
+            
+            // Chain is presented by peer 1.
+            result = cht.ConnectNewHeaders(1, listOfCurrentChainHeaders);
+
+            // Headers h5-h9 are marked as "assumed valid".
+            ChainedHeader consumed = result.Consumed;
+            var expectedState = ValidationState.Unknown;
```

Comment:
I believe this should be header only instead of unknown.

Reply:
I get Unknown returned. Bug?

Reply:
Probably, will check that. From the top of the head it shouldn't be unknown.

Reply:
This has been fixed in the latest master